addons/bus/static/src/js/web_client_bus.js

```diff
@@ -0,0 +1,92 @@
+odoo.define('bus.WebClient', function (require) {
+"use strict";
+
+var core = require('web.core');
+var WebClient = require('web.WebClient');
+
+var _t = core._t;
+
+WebClient.include({
+    /**
+     * Detects the presence of assets in DOM's HEAD
+     *
+     * @override
+     */
+    start: function () {
+        var self = this;
+        this.assetsChangedNotificationId = null;
+        this._assets = {};
+
+        return this._super.apply(this, arguments).then(function () {
+            var $assets = $('*[data-asset-xmlid]');
+            $assets.each(function () {
+                self._assets[$(this).data('asset-xmlid')] = $(this).data('asset-version');
+            });
+        });
+    },
+    /**
+     * Assign handler to bus notification
+     *
+     * @override
+     */
+    show_application: function () {
+        this.call('bus_service', 'onNotification', this, this._onNotification);
+        this.call('bus_service', 'startPolling');
+        return this._super.apply(this, arguments);
+    },
+    /**
+     * Displays one notification on user's screen when assets have changed
+     *
+     * @private
+     */
+    _displayBundleChangedNotification: function () {
+        var self = this;
+        if (!this.assetsChangedNotificationId) {
+            this.assetsChangedNotificationId = this.call('notification', 'notify', {
+                title: _t('Update available'),
+                message: _t('Refresh your browser to take advantage of the latest update of Odoo in your browser'),
+                sticky: true,
+                onClose: function () {
+                    self.assetsChangedNotificationId = null;
+                },
+                buttons: [{
+                    text: _t('Refresh'),
+                    primary: true,
+                    click: function () {
+                        window.location.reload(true);
+                    }
+                }],
+             });
+        }
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * Reacts to bus's notification
+     *
+     * @private
+     * @param {Array} notifications: list of received notifications
+     */
+    _onNotification: function (notifications) {
+        var seenXmlIds = [];
+        for (var i in notifications) {
+            var notif = notifications[i];
+            if (notif[0][1] === 'bundle_changed') {
+                var bundleXmlId = notif[1][0];
+                var bundleVersion = notif[1][1];
+                if (!seenXmlIds.includes(bundleXmlId)) {
```

Comment:
what's the purpose of seenXmlIds?

Reply:
notifications contains only one of either JS or CSS asset changed but with the same xml id
So potentially, when both CSS and JS of asset_backend have changed, you'd have two notifs for one XML ID, containing with empirically the same hash
To avoid going many times in the loop, seen xml id is necessary, although it might be kinda overkill since the notification that is displayed is a singleton

Reply:
you break as soon as you have a notification to display, so it's indeed overkill I think